pkg/util/docker/dockerfile/dockerfile_test.go

```diff
@@ -371,7 +371,60 @@ EXPOSE 9090 9091
 			t.Errorf("%s: parse error: %v", name, err)
 			continue
 		}
-		got := exposedPorts(node.AST)
+		got, _ := exposedPorts(node.AST)
+		if !reflect.DeepEqual(got, tc.want) {
+			t.Errorf("exposedPorts: %s: got %#v; want %#v", name, got, tc.want)
+		}
+	}
+}
+
+// TestLastExposedPortsVar tests calling exposedPorts with ENV,ARG and range
+func TestLastExposedPortsVarAndRange(t *testing.T) {
+	testCases := map[string]struct {
+		in   string
+		want []string
+	}{
+		"EXPOSE with ENV": {
+			in: `FROM centos:7
+ENV PORT 8080
+EXPOSE $PORT`,
+			want: []string{"8080"},
+		},
+		"EXPOSE with RANGE": {
+			in: `FROM centos:7
+EXPOSE 8080-9000`,
+			want: []string{"8080"},
+		},
+		"EXPOSE with default ARG": {
+			in: `FROM centos:7
+ARG PORT=8080
+EXPOSE $PORT`,
+			want: []string{"8080"},
+		},
+		"EXPOSE recursive ARGS, ENV and range": {
+			in: `FROM centos:7
+ARG PORT=8080-8085
+ENV PORT2 "$PORT"
+ENV PORT3="$PORT2" PORT4=$PORT3
+ENV PORT5=$PORT4
+EXPOSE $PORT5`,
+			want: []string{"8080"},
+		},
+		"EXPOSE ignore post eval": {
+			in: `FROM centos:7
+ENV PORT=8080
+EXPOSE $PORT
+ENV PORT=10`,
+			want: []string{"8080"},
+		},
```

Comment:
please add tests to cover a dockerfile like:

```
FROM foo
ENV p=8080
FROM bar
ENV p=8081
EXPOSE $p
```
(should expose 8081, not 8080)

and
```
FROM foo
ENV p=8080
FROM bar
EXPOSE $p
```
(should be an error I think, $p is not defined in the relevant FROM stanza)

Reply:
thanks, will take a look at that. I am also missing a bunch of functionality as defined in

https://docs.docker.com/engine/reference/builder/#env
https://docs.docker.com/engine/reference/builder/#environment-replacement

I will try to reuse the `shell_parser.go` from vendored dependency https://github.com/docker/docker instead so we have consistent behavior

https://github.com/openshift/origin/blob/master/vendor/github.com/docker/docker/builder/dockerfile/shell_parser.go

Reply:
I added both tests and tried to rewrite using `shell_parser.go` to fix some other issues, but I struggled with importing entire `github.com/docker/docker/builder/dockerfile` package. It looks like there may be a package conflict somewhere in our dependencies, this is what glide tells me when trying only to import the package
```
[ERROR] Error scanning github.com/containerd/containerd/api/grpc/types: cannot find package "." in:
        .../.glide/cache/src/https-github.com-containerd-containerd/api/grpc/types
```

so for now, I copied the file over, but I would like to ask for an advice how to resolve this properly. The path in `containerd/containerd/api/grpc` is probably from an older version of [containerd](https://github.com/containerd/containerd/tree/v0.2.x/api), should I try to get newer `github.com/docker/docker` in our vendor?

Reply:
> should I try to get newer github.com/docker/docker in our vendor?

no, we're tied to the level of docker that k8s uses upstream.

can you not import docker/docker/builder/dockerfile at the level we're already using for docker/docker?  It looks like we already have the docker/builder/dockerfile package vendored?

Reply:
ack, I have tried to pin the `github.com/containerd/containerd/api/grpc/types` to `v0.2.x` instead, because that appears to be needed by our version of `github.com/docker/docker/builder/dockerfile` but that breaks dependencies in `mattn/go-shellwords`
```
[INFO]  --> Setting version for github.com/mattn/go-shellwords to 02e3cf038dcea8290e44424da473dd12be796a8a.
[ERROR] Error scanning github.com/containerd/containerd/api/services/containers/v1: cannot find package "." in:
        .../.glide/cache/src/https-github.com-containerd-containerd/api/services/containers/v1
[ERROR] Error scanning github.com/containerd/containerd/api/services/tasks/v1: cannot find package "." in:
        .../.glide/cache/src/https-github.com-containerd-containerd/api/services/tasks/v1
[ERROR] Error scanning github.com/containerd/containerd/api/services/version/v1: cannot find package "." in:
        .../.glide/cache/src/https-github.com-containerd-containerd/api/services/version/v1
[ERROR] Error scanning github.com/containerd/containerd/containers: cannot find package "." in:
        .../.glide/cache/src/https-github.com-containerd-containerd/containers
[ERROR] Error scanning github.com/containerd/containerd/dialer: cannot find package "." in:
        .../.glide/cache/src/https-github.com-containerd-containerd/dialer
[ERROR] Error scanning github.com/containerd/containerd/errdefs: cannot find package "." in:
        .../.glide/cache/src/https-github.com-containerd-containerd/errdefs
[ERROR] Error scanning github.com/containerd/containerd/namespaces: cannot find package "." in:
        .../.glide/cache/src/https-github.com-containerd-containerd/namespaces
```
Unfortunatelly, I don't have any other ideas apart from including the file `shell_parser.go` as done in this PR

Reply:
ok i think what you've done is probably the path of least resistance, let's just leave a TODO to come back and remove the copied code once our dependencies allow for it.